File: src/_cpp/TowerDefenseBase/scenes/gamegs/MenuCreateTower.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "ml/Language.h"
#include "MenuCreateTower.h"
#include "GameGS.h"
#include "ScoreCounter.h"
#include "Tower.h"
#include "support.h"
#include "UserData.h"
#include "Tutorial.h"
#include "configuration.h"

NS_CC_BEGIN

MenuCreateTower::MenuCreateTower()
: _disabled( true )
{}

MenuCreateTower::~MenuCreateTower()
{
	ScoreCounter::shared().observer( kScoreLevel ).remove( _ID );
}

bool MenuCreateTower::init()
{
	ScrollMenu::init();
	NodeExt::init();

	NodeExt::load( "ini/gamescene", "menucreatetower.xml" );

	unsigned size = getItemsCount();
	for( unsigned i = 0; i < size; ++i )
	{
		auto item = getItem( i );
		if( item->getName() == "confirm" )
		{
			_confirmButton = item;
			_confirmButton->setVisible( false );
			item->setCallback( std::bind( &MenuCreateTower::confirmSelect, this, std::placeholders::_1, true ) );
		}
		else if( item->getName() == "confirm_un" )
		{
			_confirmButtonUn = item;
			_confirmButtonUn->setVisible( false );
			item->setCallback( std::bind( &MenuCreateTower::confirmSelect, this, std::placeholders::_1, false ) );
		}
		else
		{
			std::string name = item->getName( );
			if( name.find( "_un" ) == std::string::npos )
			{
				_buttonTowers[name] = item;
				item->setCallback( std::bind( &MenuCreateTower::onActivate, this, std::placeholders::_1, true ) );
			}
			else
			{
				name.erase( name.begin() + (name.size() - 3), name.end() );
				_buttonTowersUn[name] = item;
				item->setCallback( std::bind( &MenuCreateTower::onActivate, this, std::placeholders::_1, false ) );
			}
			int level = UserData::shared().tower_upgradeLevel( name );
			if( level == 0 )
				setBclokButton( item );
		}
	}

	_desc.node = getChildByName( "desc" );
	if( _desc.node )
	{
		_desc.name.reset( _desc.node->getChildByName<Label*>( "name" ) );
		_desc.text.reset( _desc.node->getChildByName<Label*>( "text" ) );
		_desc.dmg.reset( _desc.node->getChildByName<Label*>( "dmg" ) );
		_desc.rng.reset( _desc.node->getChildByName<Label*>( "rng" ) );
		_desc.spd.reset( _desc.node->getChildByName<Label*>( "spd" ) );
	}

	setVisible( false );
	return true;
}

void MenuCreateTower::setBclokButton( MenuItem*button )
{
	button->setCallback( CC_CALLBACK_1( MenuCreateTower::onBlocked, this ) );
	auto item = dynamic_cast<MenuItemImageWithText*>(button);
	if( item )
	{
		item->setImageNormal( k::resourceGameSceneFolder + "menucreatetower/menu_lock2.png" );
		item->setImageSelected( k::resourceGameSceneFolder + "menucreatetower/menu_lock2.png" );
	}
}

void MenuCreateTower::onBlocked( Ref * sender )
{
	disappearance();
}

void MenuCreateTower::onActivate( Ref * sender, bool availebledButton )
{
	if( _disabled )return;
	hideConfirmButton();

	auto node = dynamic_cast<MenuItem*>(sender);
	if( !node ) return;
	_hidenButton.reset( node );
	//_hidenButton->setVisible( false );

	std::string name = _hidenButton->getName();

	if( availebledButton )
	{
		_confirmButton->setPosition( node->getPosition() );
		_confirmButton->setVisible( true );
	}
	else
	{
		name.erase( name.begin() + name.size() - 3, name.end() );
		_confirmButtonUn->setName( name );
		_confirmButtonUn->setPosition( node->getPosition() );
		_confirmButtonUn->setVisible( true );
	}

	_selectedTower = name;

	runEvent( "onclick" );
	runEvent( "onclickby_" + _selectedTower );
	buildDescription();

	//float rate = mlTowersInfo::shared( ).rate( name );
	float radius = mlTowersInfo::shared().radiusInPixels( name, 1 );
	showRadius( _centerPoint, radius /** rate*/ );
}

void MenuCreateTower::confirmSelect( Ref * sender, bool availebledButton )
{
	if( _disabled )return;
	if( availebledButton )
	{
		assert( _hidenButton );
		std::string towername = _hidenButton->getName();
		if(towername.find("_un") != std::string::npos )
		{
			towername = towername.substr( 0, towername.find("_un") );
		}
		hideConfirmButton();
		runEvent( "onconfirm" );

		GameGS::getInstance()->getGameBoard().createTower( towername );
		disappearance();
	}
	else
	{
		if(k::configuration::useInapps && TutorialManager::shared().dispatch( "level_haventgear_build" ) )
			disappearance();
	}
}

void MenuCreateTower::hideConfirmButton()
{
	_confirmButton->setVisible( false );
	_confirmButtonUn->setVisible( false );
	if( _hidenButton )
		_hidenButton->setVisible( true );
	_hidenButton.reset( nullptr );
}

void MenuCreateTower::changeCost()
{
	auto set = []( Node*node, int cost )
	{
		auto label = node->getChildByName<Label*>( "cost" );
		if( label ) label->setString( intToStr( cost ) );
	};

	for( auto& icon : _buttonTowers )
	{
		unsigned cost = mlTowersInfo::shared().getCost( icon.first, 1 );
		cost = std::min<unsigned>( cost, 999 );
		auto& but0 = icon.second;
		auto& but1 = _buttonTowersUn[icon.first];
		set( but0, cost );
		set( but1, cost );
	}

}

void MenuCreateTower::appearance()
{
	setVisible( true );
	hideConfirmButton();
	runEvent( "appearance" );
	_disabled = false;
	ScoreCounter::shared().observer( kScoreLevel ).add( _ID, std::bind( &MenuCreateTower::onChangeMoney, this, std::placeholders::_1 ) );
	changeCost();
	onChangeMoney( ScoreCounter::shared().getMoney( kScoreLevel ) );
	scheduleUpdate();
}

void MenuCreateTower::disappearance()
{
	if( _disabled == false )
	{
		hideConfirmButton();
		runEvent( "disappearance" );
		_disabled = true;

		hideRadius();
		ScoreCounter::shared().observer( kScoreLevel ).remove( _ID );
		unscheduleUpdate();
	}
}

void MenuCreateTower::setExcludedTowers( const std::list<std::string> & list ) {}

void MenuCreateTower::addExludedTower( const std::string & towerName ) 
{
	auto item = getItemByName( towerName );
	if( item )
		setBclokButton( item );
}

void MenuCreateTower::removeExludedTower( const std::string & towerName ) 
{
}

void MenuCreateTower::setActived( bool mode ) {}
void MenuCreateTower::setClickPoint( const Point & point )
{
	_centerPoint = point;
	setPosition( point );
}

void MenuCreateTower::buildDescription()
{
	std::string name = _selectedTower;
	name = WORD( name + "_name" );
	std::string dmg = intToStr( mlTowersInfo::shared().get_dmg( _selectedTower, 1 ) );
	std::string rng = intToStr( mlTowersInfo::shared().get_rng( _selectedTower, 1 ) );
	std::string spd = intToStr( mlTowersInfo::shared().get_spd( _selectedTower, 1 ) );
	std::string txt = mlTowersInfo::shared().get_desc( _selectedTower, 1 );
	if( _desc.name )_desc.name->setString( name );
	if( _desc.dmg )_desc.dmg->setString( dmg );
	if( _desc.rng )_desc.rng->setString( rng );
	if( _desc.spd )_desc.spd->setString( spd );
	if( _desc.text )_desc.text->setString( txt );
}

void MenuCreateTower::onChangeMoney( int money )
{
	for( auto& iter : _buttonTowers )
	{
		auto& but0 = iter.second;
		auto& but1 = _buttonTowersUn[iter.first];
		assert( but0 && but1 );

		int cost = mlTowersInfo::shared().getCost( iter.first, 1 );

		but0->setVisible( cost <= money );
		but1->setVisible( cost > money );
	}

	//	if( _hidenButton)
	//		_hidenButton->setVisible( false );

	if( _confirmButtonUn->isVisible() )
	{
		int cost = mlTowersInfo::shared().getCost( _confirmButtonUn->getName(), 1 );
		if( cost <= money )
		{
			_confirmButtonUn->setVisible( false );
			_confirmButton->setVisible( true );
			_confirmButton->setPosition( _confirmButtonUn->getPosition() );
		}
	}
}

void MenuCreateTower::update( float )
{
	auto point = GameGS::getInstance()->getMainLayer()->convertToWorldSpace( _centerPoint );
	setPosition( point );

	if( _desc.node )
	{
		auto screenpos = getPosition();
		auto dessize = Director::getInstance( )->getOpenGLView( )->getDesignResolutionSize( );

		auto pos = _desc.node->getPosition( );
		pos.x = fabs( pos.x );
		_desc.node->setAnchorPoint( Point( 0, 0.5f ) );
		if( screenpos.x > dessize.width / 2 )
		{
			pos.x = -pos.x;
			_desc.node->setAnchorPoint( Point( 1, 0.5f ) );
		}
		_desc.node->setPosition( pos );
	}
}

NS_CC_END

File: src/_eu/scenes/gamegs/MenuCreateTower.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};


EU.MenuCreateTower = EU.ScrollMenu.extend(
{
    __MenuCreateTower : true,

    /** @type {bool} */ _disabled : true,
    /** @type {string} */ _selectedTower : "" ,
    /** @type {cc.MenuItem} */ _confirmButton : null,
    /** @type {cc.MenuItem} */ _confirmButtonUn : null,
    /** @type {cc.MenuItem} */ _hidenButton : null,
    _desc : null,

    /** @type {std.map< std.string, MenuItem >} */ _buttonTowers : null,
    /** @type {std.map< std.string, MenuItem >} */ _buttonTowersUn : null,
    /** @type {cc.Point} */ _centerPoint : null,

    onExit: function()
    {
        EU.ScoreCounter.observer( EU.kScoreLevel ).remove( this.__instanceId );
        this._super();
    },
    ctor: function()
    {
        this._super();
        this.initExt();
        this._buttonTowers = {};
        this._buttonTowersUn = {};
        this._centerPoint = cc.p(0,0);
        this._desc = {
            /**cc.Node*/ node: new cc.Node(),
            /**cc.LabelTTF*/ name: new cc.LabelTTF(),
            /**cc.LabelTTF*/ text: new cc.LabelTTF(),
            /**cc.LabelTTF*/ dmg: new cc.LabelTTF(),
            /**cc.LabelTTF*/ rng: new cc.LabelTTF(),
            /**cc.LabelTTF*/ spd: new cc.LabelTTF()
        };

        this.load_str_n_str( "ini/gamescene", "menucreatetower.xml" );

        var self = this;

        var size = this.getItemsCount();
        for( var i = 0; i < size; ++i )
        {
            var item = this.getItem( i );
            if( item.getName() == "confirm" )
            {
                this._confirmButton = item;
                this._confirmButton.setVisible( false );
                this._confirmButton.setCallback(this.confirmSelect.bind(this,true), this);
            }
            else if( item.getName() == "confirm_un" )
            {
                this._confirmButtonUn = item;
                this._confirmButtonUn.setVisible( false );
                this._confirmButtonUn.setCallback( this.confirmSelect.bind(this,true), this);
            }
            else
            {
                var name = item.getName( );
                if( name.indexOf( "_un" ) < 0 )
                {
                    this._buttonTowers[name] = item;
                    this._buttonTowers[name].setCallback( this.onActivate.bind(this, item, true), this );
                }
                else
                {
                    //name.erase( name.begin() + (name.length - 3), name.end() );
                    name = name.slice( 0, -3);

                    this._buttonTowersUn[name] = item;
                    this._buttonTowersUn[name].setCallback( this.onActivate.bind(this, item, false), this );
                }
                var level = EU.UserData.tower_getUpgradeLevel( name );
                if( level == 0 )
                    this.setBlockButton( item );
            }
        }

        this._desc.node = this.getChildByName( "desc" );
        if( this._desc.node )
        {
            this._desc.name = ( this._desc.node.getChildByName( "name" ) );
            this._desc.text = ( this._desc.node.getChildByName( "text" ) );
            this._desc.dmg = ( this._desc.node.getChildByName( "dmg" ) );
            this._desc.rng = ( this._desc.node.getChildByName( "rng" ) );
            this._desc.spd = ( this._desc.node.getChildByName( "spd" ) );
        }

        this.setVisible( false );
        return true;
    },

    setBlockButton: function( /**cc.MenuItem*/ button )
    {
        var self = this;
        button.setCallback( this.onBlocked.bind(this));
        var item = button;
        if( item.__MenuItemImageWithText)
        {
            item.setImageNormal( EU.k.resourceGameSceneFolder + "menucreatetower/menu_lock2.png" );
            item.setImageSelected( EU.k.resourceGameSceneFolder + "menucreatetower/menu_lock2.png" );
        }
    },

    onBlocked: function(  sender )
    {
        this.disappearance();
    },

    onActivate: function(  sender, availableButton )
    {
        if( this._disabled )return;
        this.hideConfirmButton();

        var node = sender;
        if( ! (node instanceof cc.MenuItem)) return;
        //TODO: Implement InstrusivePtr on cc.Class.prototype
        this._hidenButton = ( node );
        //this._hidenButton.setVisible( false );

        var name = this._hidenButton.getName();

        if( availableButton )
        {
            this._confirmButton.setPosition( node.getPosition() );
            this._confirmButton.setVisible( true );
            this._buttonTowers[name].setVisible( false );
        }
        else
        {
            name = name.substring(0, name.length - 3);
            this._confirmButtonUn.setName( name );
            this._confirmButtonUn.setPosition( node.getPosition() );
            this._confirmButtonUn.setVisible( true );
        }

        this._selectedTower = name;

        this.runEvent( "onclick" );
        this.runEvent( "onclickby_" + this._selectedTower );
        this.buildDescription();

        //var rate = EU.mlTowersInfo.shared( ).rate( name );
        var radius = EU.mlTowersInfo.radiusInPixels( name, 1 );
        EU.showRadius( this._centerPoint, radius /** rate*/ );
    },

    confirmSelect: function(  sender, availableButton )
    {
        if( this._disabled )return;
        if( availableButton )
        {
            EU.assert( this._hidenButton );
            var towername = this._hidenButton.getName();
            if(towername.indexOf("_un") >= 0 )
            {
                towername = towername.substr( 0, towername.indexOf("_un") );
            }
            this.hideConfirmButton();
            this.runEvent( "onconfirm" );

            EU.GameGSInstance.getGameBoard().createTower( towername );
            this.disappearance();
        }
        else
        {
            if(EU.k.configuration.useInapps && EU.TutorialManager.dispatch( "level_haventgear_build" ) )
                this.disappearance();
        }
    },

    hideConfirmButton: function()
    {
        this._confirmButton.setVisible( false );
        this._confirmButtonUn.setVisible( false );
        if( this._hidenButton )
            this._hidenButton.setVisible( true );
        this._hidenButton = ( null );
    },

    changeCost: function()
    {
        var set = function( node, cost )
        {
            var label = node.getChildByName( "cost" );
            if( label ) label.setString( cost);
        };

        for( var icon in this._buttonTowers )
        {
            if (this._buttonTowers.hasOwnProperty(icon)) {
                var cost = EU.mlTowersInfo.getCost( icon, 1 );
                cost = Math.min( cost, 999 );
                var but0 = this._buttonTowers[icon];
                var but1 = this._buttonTowersUn[icon];
                set( but0, cost );
                set( but1, cost );
            }
        }

    },

    appearance: function()
    {
        this.setVisible( true );
        this.hideConfirmButton();
        this.runEvent( "appearance" );
        this.setEnabled( true );
        this._disabled = false;
        EU.ScoreCounter.observer( EU.kScoreLevel ).add( this.__instanceId, this.onChangeMoney, this);
        this.changeCost();
        this.onChangeMoney( EU.ScoreCounter.getMoney( EU.kScoreLevel ) );
        this.scheduleUpdate();
    },
    disappearance: function()
    {
        if( this._disabled == false )
        {
            this.hideConfirmButton();
            this.runEvent( "disappearance" );
            this.setEnabled( false );
            this._disabled = true;

            EU.hideRadius();
            EU.ScoreCounter.observer( EU.kScoreLevel ).remove( this.__instanceId );
            this.unscheduleUpdate();
        }
    },

    setExcludedTowers: function( list ) {},

    addExludedTower: function( towerName )
    {
        var item = this.getItemByName( towerName );
        if( item )
            this.setBlockButton( item );
    },

    removeExludedTower: function(  towerName )
    {
    },

    setActived: function( mode ) {},
    setClickPoint: function(  point )
    {
        this._centerPoint = point;
        this.setPosition( point );
    },

    buildDescription: function()
    {
        var name = this._selectedTower;
        name = EU.Language.string( name + "_name" );
        var dmg = ( EU.mlTowersInfo.get_dmg( this._selectedTower, 1 ) );
        var rng = ( EU.mlTowersInfo.get_rng( this._selectedTower, 1 ) );
        var spd = ( EU.mlTowersInfo.get_spd( this._selectedTower, 1 ) );
        var txt = EU.mlTowersInfo.get_desc( this._selectedTower, 1 );
        if( this._desc.name )this._desc.name.setString( name );
        if( this._desc.dmg )this._desc.dmg.setString( dmg );
        if( this._desc.rng )this._desc.rng.setString( rng );
        if( this._desc.spd )this._desc.spd.setString( spd );
        if( this._desc.text )this._desc.text.setString( txt );
    },

    onChangeMoney: function( money )
    {
        for( var iter in this._buttonTowers )
        if (this._buttonTowers.hasOwnProperty(iter)) {

            var but0 = this._buttonTowers[iter];
            var but1 = this._buttonTowersUn[iter];
            EU.assert( but0 && but1 );

            var cost = EU.mlTowersInfo.getCost( iter, 1 );

            but0.setVisible( cost <= money );
            but1.setVisible( cost > money );
        }

        //	if( this._hidenButton)
        //		this._hidenButton.setVisible( false );

        if( this._confirmButtonUn.isVisible() )
        {
            cost = EU.mlTowersInfo.getCost( this._confirmButtonUn.getName(), 1 );
            if( cost <= money )
            {
                this._confirmButtonUn.setVisible( false );
                this._confirmButton.setVisible( true );
                this._confirmButton.setPosition( this._confirmButtonUn.getPosition() );
            }
        }
    },

    update: function( )
    {
        var point = EU.GameGSInstance.getMainLayer().convertToWorldSpace( this._centerPoint );
        this.setPosition( point );

        if( this._desc.node )
        {
            var screenpos = this.getPosition();
            var dessize = cc.view.getDesignResolutionSize();

            var pos = this._desc.node.getPosition( );
            pos.x = Math.abs( pos.x );
            this._desc.node.setAnchorPoint( cc.p( 0, 0.5 ) );
            if( screenpos.x > dessize.width / 2 )
            {
                pos.x = -pos.x;
                this._desc.node.setAnchorPoint( cc.p( 1, 0.5 ) );
            }
            this._desc.node.setPosition( pos );
        }
    }

});

EU.NodeExt.call(EU.MenuCreateTower.prototype);
